### PR TITLE
Pull request for review

### PR DESCRIPTION
**Pull request by napafundi**
initial setup commit, added ember-cli-typescript, and prettier support. Updated readme and other config options

### CODE REVIEW

File: addon/components/row/loading.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+export default class RowLoadingComponent extends Component {
+    layout = layout;
+    tagName: string = 'tr';
+    classNames: string[] = ['et-tr', 'data-table-row-loading'];
+    cellClassNames: string = 'p-4';
+
+    //readonly attributes
+
+    @readOnly('columns.length') colspan?: number;
```
Comment (bakerac4) [inline addon/components/row/loading.ts]: @napafundi why would `colspan` be optional?

File: addon/components/gavant-table.ts [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+
+    /**
+     * Invokes an action to load a new page of rows
+     *
+     * @returns {Promise}
+     * @memberof GavantTableComponent
+     */
+    @action
+    onLastReached() {
+        if (!this.isLoading && this.hasMoreRows) {
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: I wouldn't use `tryInvoke` as Ember is deprecating most of the `try` helpers and it sorta ruins the type safety.

Right now your comments on this action are possibly incorrect, because loadMoreRows might not exist and therefore it returns the function that tryInvoke returns (not a Promise). I think you will want to check if `this.loadMoreRows` exists in the if check, and then just call it

File: addon/components/gavant-table.ts [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+
+    /**
+     * Updates the table's column sorting
+     *
+     * @param {any[]} sorts
+     * @returns {Promise}
+     * @memberof GavantTableComponent
+     */
+    @action
+    onUpdateSorts(sorts: any[]) {
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: What is a `sort`? Is there a certain shape that it should follow?

File: addon/components/gavant-table.ts [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+    }
+
+    /**
+     * Returns the column's unique identifier or undefined
+     *
+     * @param {(TableColumn | undefined)} col - A table column
+     * @returns {(string | undefined)}
+     * @memberof GavantTableComponent
+     */
+    getColumnId(col: TableColumn | undefined): string | undefined {
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: You can change this to `col?: TableColumn` which means that col might be not passed in and undefined

File: addon/components/gavant-table.ts [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+    }
+
+    /**
+     * Returns the computed width for the given element
+     *
+     * @param {Element} el -- A DOM element
+     * @returns {number}
+     * @memberof GavantTableComponent
+     */
+    getElementWidth(el: Element): number {
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: This should probably be either A) made private, or b) made into a helper function that's imported by the table

File: addon/components/gavant-table.ts [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+    }
+
+    /**
+     * Handle component container element resize events and hide/show columns as needed
+     *
+     * @memberof GavantTableComponent
+     */
+    debouncedDidResize() {
+        if (this.constrainColumnsToFit) {
+            //reset pan position on container resize
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: Should probably move these comments into the main comment section of the method so they are visible

File: addon/components/gavant-table.ts [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+        } else {
+            return this.fillColumnIndex;
+        }
+    }
+
+    /**
+     * Updates the visible columns when the source columns array changes
+     *
+     * @readonly
+     * @type {*} any
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: Is this the correct type?

File: addon/components/row/loading.ts (hunk 1 above)
Comment (napafundi) [reply addon/components/row/loading.ts]: Wasn't sure how to properly handle these types of values in TS. There are two options recommended. 1) adding `| undefined` to the declaration, and 2) adding `!:` definite declaration. Since the columns value is guaranteed, I'm guessing the definite declaration is the way to go?

File: addon/components/gavant-table.ts [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+     * CP that returns a new fillColumnIndex when the current
+     * fillColumnIndex is out of range
+     *
+     * @readonly
+     * @type {(number | null)} adjustedFillColumnIndex
+     * @memberof GavantTableComponent
+     */
+    @computed('fillColumnIndex', 'visibleColumns.length')
+    get adjustedFillColumnIndex(): number | null {
+        if (this.fillColumnIndex) {
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: I find this statement sort of confusing. If `this object` else return `this object`. Might not hurt to just return null (if that's what its returning in that case)

File: addon/components/gavant-table.ts [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+     */
+    @computed(
+        'hasHiddenColumns',
+        'allowFixedCols',
+        'columns.firstObject.{id,valuePath}',
+        'nonFixedColumns.firstObject.{id,valuePath}',
+        'firstVisibleColumn.{id,valuePath}'
+    )
+    get canPanLeft(): boolean {
+        //columns can be panned left if there are hidden columns
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: Probably should move these comments up

File: addon/components/gavant-table.ts [hunk 10]
```diff
@@ -0,0 +1,532 @@
...
+
+    /**
+     * CP that returns which column should be used as the first
+     * visible column
+     *
+     * @type {TableColumn} firstVisibleColumn
+     * @memberof GavantTableComponent
+     */
+    @conditional('allowFixedCols', 'firstVisibleNonFixedColumn', 'visibleColumns.firstObject')
+    firstVisibleColumn!: TableColumn;
```
Comment (bakerac4) [inline addon/components/gavant-table.ts]: Is it possible for this object to not be defined? A table with 0 columns? Or do we prevent that elsewhere?

File: addon/components/gavant-table.ts (hunk 2 above)
Comment (napafundi) [reply addon/components/gavant-table.ts]: Makes sense. Since this will require declaring a property called `loadMoreRows`, what's the preferred type for functions?

**Comment by bakerac4** [comment]
@napafundi nice work. Overall it looks good. 2 general things is we need to make some of the methods in that table private so think about what a person should be using and what should be private and not used.  I also noticed you put more detailed comments inside a method, which means they won't be visible in the popover window. If they have anything to do with how to use that method, I'd recommend bringing those out.

File: addon/components/row/loading.ts (hunk 1 above)
Comment (billdami) [reply addon/components/row/loading.ts]: yeah i think we've gone with the `!` approach in other cases like this before (such as when defining types for injected services), because we know that that computed macro will always return a value.

File: addon/components/gavant-table.ts (hunk 8 above)
Comment (billdami) [reply addon/components/gavant-table.ts]: I am pretty sure we are handling the non-zero case here, so this might be safe to change to `if (this.fillColumnIndex > 0) { ...` to be more descriptive. but i would double-check that @napafundi

File: addon/components/gavant-table.ts (hunk 2 above)
Comment (billdami) [reply addon/components/gavant-table.ts]: @napafundi depends on what params the function accept, and what its return type is, but in general you'll define function types like this:
```ts
someFunc: (param1: type, param2: anotherType) => returnType;
```
in this case, I think this action takes no params and returns a promise, so probably something like:
```ts
loadMoreRows?: () => Promise<any[]>;
```
which means an optional function with no arguments that returns a promise (if it returned nothing, that would be `void` instead of `Promise`) which resolves with an array of objects in any shape (that could possibly be more specific, but the table doesn't really care about the row shapes, so I think thats probably ok in this case).

File: addon/components/gavant-table.ts (hunk 10 above)
Comment (billdami) [reply addon/components/gavant-table.ts]: i dont remember that being actively prevented anywhere, so maybe to be safe we should take out this `!`, and then add if blocks/guards in places where we are assuming this to have a value.

File: .template-lintrc.js [hunk 11]
```diff
@@ -1,5 +0,0 @@
-'use strict';
-
-module.exports = {
-  extends: 'octane'
-};
```
Comment (billdami) [inline .template-lintrc.js]: did we have to remove the template linting config for a specific issue?

File: README.md [hunk 12]
```diff
@@ -1,38 +1,27 @@
...
 ```
 ember install gavant-ember-table
 ```
 
-
-Usage
-------------------------------------------------------------------------------
+## Usage
 
 [Longer description of how to use the addon in apps.]
```
Comment (billdami) [inline README.md]: not super high priority, but I would add at least a basic example of how to use the table, and then probably just provide a link to the official ember-table docs website, since 99% of that applicable here. You can check out some of our other more recent addons for examples on how we wrote usage guides.

File: package.json [hunk 13]
```diff
@@ -1,63 +1,93 @@
 {
-  "name": "gavant-ember-table",
```
Comment (billdami) [inline package.json]: this should be changed to `@gavant/ember-table` before we publish

File: README.md [hunk 14]
```diff
@@ -1,38 +1,27 @@
...
-* Ember CLI v2.13 or above
-* Node.js v8 or above
-
-
-Installation
-------------------------------------------------------------------------------
+## Installation
 
 ```
 ember install gavant-ember-table
```
Comment (billdami) [inline README.md]: I think this needs to be changed to `ember install @gavant/ember-table`

File: addon/components/gavant-table.ts [hunk 15]
```diff
@@ -0,0 +1,532 @@
...
+import { lt, gte, conditional } from 'ember-awesome-macros';
+import ResizeAware from 'ember-resize/mixins/resize-aware';
+import Media from 'ember-responsive';
+import { observes } from '@ember-decorators/object';
+import NativeArray from '@ember/array/-private/native-array';
+// @ts-ignore: Ignore import of compiled template
+import layout from '../templates/components/gavant-table';
+import { TableColumn } from 'table-components';
+
+class GavantTableComponent extends Component.extend({ ResizeAware }) {
```
Comment (billdami) [inline addon/components/gavant-table.ts]: If we can I'd probably prefer we keep this component named just `TableComponent`, we generally haven't prefixed any of our components in other addons with `Gavant`. And the single word `<Table>` component instantiation is just so clean :-D

File: addon/components/gavant-table.ts [hunk 16]
```diff
@@ -0,0 +1,532 @@
...
+    //methods
+
+    /**
+     * Initialize the table component's `visibleColumns`
+     *
+     * @memberof GavantTableComponent
+     */
+    init() {
+        super.init();
+        this.columns.forEach((col) => {
```
Comment (billdami) [inline addon/components/gavant-table.ts]: you could probably shorten this to
```ts
this.columns.setEach('isVisible', true);
```

File: addon/components/gavant-table.ts [hunk 17]
```diff
@@ -0,0 +1,532 @@
...
+
+    /**
+     * Updates the visible columns based on `visibleColumns`
+     * (an array which includes columns based on their ability
+     * to fit within the table container)
+     *
+     * @memberof GavantTableComponent
+     */
+    @action
+    changeDisplay() {
```
Comment (billdami) [inline addon/components/gavant-table.ts]: we should prob move this logic directly into the `updateColumnVisibility` method since its not really an action that is intended to be called anywhere else outside of that. and we might be able to get away with not having an extra loop on the columns aray/having to find them by valuePath, if we just do it directly in the updateColumnVisibility() (but be sure to comment all the affected lines when we move it there, so we know what to rip out once the real bug is fixed)

File: addon/templates/components/gavant-table.hbs [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+        <Head.row as |Row|>
+            <Row.cell
+                class={{concat
+                    Row.columnValue.headerClassNames
+                    (if Row.columnValue.isFixedLeft " data-table-col-fixed-left")
+                    (if (eq Row.columnValue this.firstVisibleNonFixedColumn) " data-table-col-first-nonfixed")
+                }}
+                as |columnValue columnMeta|
+            >
+                {{#if (and columnValue.headerComponent columnValue.isVisible)}}
```
Comment (billdami) [inline addon/templates/components/gavant-table.hbs]: i would probably add comments above these lines where we use the `isVisible` prop, so we know where we are using this temp fix when the bug gets fixed.

File: app/formats.js [hunk 19]
```diff
@@ -0,0 +1,30 @@
+export default {
```
Comment (billdami) [inline app/formats.js]: do we have anything in the addon that has a hard dep on the ember-intl addon? if not i'd probably just rip it out, along with all these related config files.

File: app/styles/gavant-ember-table.scss [hunk 20]
```diff
@@ -0,0 +1,220 @@
+@import 'gavant-ember-table/variables';
+@import 'ember-bootstrap/bootstrap';
+@import './progress-spinner/progress-spinner';
+
+.data-table {
+    position: relative;
+    padding-bottom: 1rem;
+    padding-left: 1.25rem;
+    padding-right: 1.25rem;
```
Comment (billdami) [inline app/styles/gavant-ember-table.scss]: i'd probably make all these padding values into overrideable variables. e.g. 
```
$gavant-table-bot-padding: 1rem !default;
//..
```

File: app/styles/gavant-ember-table/_variables.scss [hunk 21]
```diff
@@ -0,0 +1,6 @@
+$table-hover-bg-opaque: #ececec !default;
+$primary: #285a92 !default;
+$white: #fff !default;
+$table-border-color: rgba(#000, 0.075) !default;
+$font-family-base: 'Lato', sans-serif !default;
+$text-muted: lighten(#6c757d, 15%);
```
Comment (billdami) [inline app/styles/gavant-ember-table/_variables.scss]: i would prefix all your addon sass vars with like `$gavant-table-` to avoid naming collisions with the project's/other addon's variables.

File: ember-cli-build.js [hunk 22]
```diff
@@ -3,16 +3,23 @@
...
+    let app = new EmberAddon(defaults, {
+        'ember-font-awesome': {
+            includeComponent: true
+        },
 
-  /*
+        'ember-bootstrap': {
+            bootstrapVersion: 4,
+            importBootstrapCSS: false
+        }
```
Comment (billdami) [inline ember-cli-build.js]: is this bootstrap stuff just for the dummy app, or is that required for the addon to build?

File: app/styles/gavant-ember-table.scss [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+            &.is-ascending:before {
+                transform: rotate(180deg);
+            }
+        }
+
+        td.is-fixed-left,
+        td.is-fixed-right,
+        th.is-fixed-left,
+        th.is-fixed-right {
+            background-color: $white;
```
Comment (billdami) [inline app/styles/gavant-ember-table.scss]: id make this var something more generic, like `$gavant-table-fixed-cell-bg`.

File: app/styles/gavant-ember-table/_variables.scss [hunk 24]
```diff
@@ -0,0 +1,6 @@
+$table-hover-bg-opaque: #ececec !default;
+$primary: #285a92 !default;
+$white: #fff !default;
+$table-border-color: rgba(#000, 0.075) !default;
+$font-family-base: 'Lato', sans-serif !default;
```
Comment (billdami) [inline app/styles/gavant-ember-table/_variables.scss]: I'd probably make this default value a more generic font-family (I think "Lato" is a custom font), maybe take a look at bootstrap's default font family.

File: app/styles/gavant-ember-table.scss [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+                border-top: 0;
+                border-bottom: none;
+            }
+        }
+
+        &.table-sm {
+            th,
+            td {
+                padding-left: 0.75rem;
+                padding-right: 0.75rem;
```
Comment (billdami) [inline app/styles/gavant-ember-table.scss]: would move these padding values into variables too

File: addon/components/gavant-table.ts (hunk 10 above)
Comment (napafundi) [reply addon/components/gavant-table.ts]: The `visibleColumns` array can have 0 columns technically if something goes wrong render-wise and the table-container has `width` less than the first column's `staticWidth`.

File: ember-cli-build.js (hunk 22 above)
Comment (napafundi) [reply ember-cli-build.js]: It's just for the dummy app

**Comment by napafundi** [comment]
@bakerac4 @billdami finished most of the CR updates - still left the `firstVisibleColumn` discussion up as it hasn't fully been resolved.

File: app/styles/gavant-ember-table.scss [hunk 26]
```diff
@@ -0,0 +1,220 @@
+@import 'gavant-ember-table/variables';
+@import 'ember-bootstrap/bootstrap';
```
Comment (billdami) [inline app/styles/gavant-ember-table.scss]: i don't think we want ember-bootstrap imported in the addon's styles, as our projects import bootstrap themselves (and usually directly via the main bootstrap repo). that way projects can use newer versions of bootstrap w/o us having to update this addon whenever a new version is released.

File: app/styles/gavant-ember-table.scss [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+@import './progress-spinner/progress-spinner';
+
+.data-table {
+    position: relative;
+    padding-bottom: $gavant-padding-bottom;
```
Comment (billdami) [inline app/styles/gavant-ember-table.scss]: i'd probably maybe these vars all start with the `$gavant-table-...` prefix so they are a bit more specific to what addon the are coming from (though `$gavant-table-` is a bit long for a prefix.. maybe we can shorten it to `$gav-tbl-` or something :-P)

File: translations/en-us.yaml [hunk 28]
```diff
@@ -0,0 +1 @@
+price_banner: The {product} costs {price, number, USD}
```
Comment (billdami) [inline translations/en-us.yaml]: looks like you have a couple things from ember-intl remaining that could be removed (including a few of the types files)

File: types/ember-table/index.d.ts [hunk 29]
```diff
@@ -0,0 +1,23 @@
+declare module 'ember-table/components/ember-table/component';
+declare module 'ember-table/components/ember-tbody/component';
+
+declare module 'table-components' {
```
Comment (billdami) [inline types/ember-table/index.d.ts]: was this table-components module name chosen specifically for a certain need? I wonder if we can change the module name to `@gavant/ember-table` to match the package name, as long as it doesn't mess up anything with TS regarding its ability to find the types correctly.

File: app/styles/gavant-ember-table.scss (hunk 26 above)
Comment (napafundi) [reply app/styles/gavant-ember-table.scss]: In the gavant-ember-table.scss, we have `@include media-breakpoint-up()` which comes from bootstrap, so will this cause issues when running in other apps? Should I indicate that the application uses bootstrap in the readme?

Comment (billdami) [reply app/styles/gavant-ember-table.scss]: @napafundi it looks like that 1 instead of that `media-breakpoint-up` mixin usage is for the "static-table" style which removes the min-height that we had applied as part of the DRS sitewide ui review changes? If thats the case, I'd probably just remove that `static-table` style block entirely, as that's a more per-project type style/customization.

Otherwise, we'd probably just convert that mixin usage to a raw css media query breakpoint, so that we could shed the addon's dependency on bootstrap. looks like we prob can avoid that in this case though if the above is true.

Comment (napafundi) [reply app/styles/gavant-ember-table.scss]: ahhh, yeah you're 100% correct.

**Comment by napafundi** [comment]
@billdami  second CR updates in.

**Comment by billdami** [comment]
@napafundi this is all set. from master, lets get a 0.0.1 version published to npm via the ember-cli-release command (you'll just need to be added to the @gavant npm org if you aren't yet). And then do one last test from a blank app, installing the addon from npm for reals, and then we should be g2g!
